File: typescript/pages/InventoryPage.ts
import { Page, Locator, expect, test } from '@playwright/test';

export class InventoryPage {
    readonly inventoryContainer: Locator;
    readonly backpackAddToCart: Locator;
    readonly shoppingCartBadge: Locator;
    readonly shoppingCartLink: Locator;

    constructor(private page: Page) {
        this.inventoryContainer = page.getByTestId('inventory-container');
        this.backpackAddToCart = page.getByTestId('add-to-cart-sauce-labs-backpack');
        this.shoppingCartBadge = page.getByTestId('shopping-cart-badge');
        this.shoppingCartLink = page.getByTestId('shopping-cart-link');
    }

    async verifyInventoryPage() {
        await test.step('Verify inventory page is displayed', async () => {
            await expect(
                this.page,
                'Expected the URL to contain inventory.html'
            ).toHaveURL(/inventory.html/);

            await expect(
                this.inventoryContainer,
                'Expected the inventory container to be visible'
            ).toBeVisible();
        });
    }

    async addBackpackToCart() {
        await test.step('Add Sauce Labs Backpack to cart', async () => {
            await this.backpackAddToCart.click();
        });

        await test.step('Verify cart contains one item', async () => {
            await expect(
                this.shoppingCartBadge,
                'Expected the shopping cart badge to display 1'
            ).toHaveText('1');
        });
    }

    async openCart() {
        await test.step('Open shopping cart', async () => {
            await this.shoppingCartLink.click();
        });
    }
}

File: python/pages/inventory_page.py
from playwright.sync_api import Page, expect
import allure


class InventoryPage:

    def __init__(self, page: Page):
        self.page = page
        self.item_container = page.locator(
            '[data-test="inventory-container"]'
        )
        self.shopping_cart_badge = page.locator(
            '[data-test="shopping-cart-badge"]'
        )
        self.shopping_cart_link = page.locator(
            '[data-test="shopping-cart-link"]'
        )
        self.backpack_button = page.locator(
            '[data-test="add-to-cart-sauce-labs-backpack"]'
        )

    def verify_inventory_page(self):
        with allure.step("Verify inventory page is displayed"):
            expect(self.page).to_have_url(
                "https://www.saucedemo.com/inventory.html"
            )
            expect(self.item_container).to_be_visible()

    def add_backpack_to_cart(self):
        with allure.step("Add Sauce Labs Backpack to cart"):
            self.backpack_button.click()

        with allure.step("Verify cart contains one item"):
            expect(self.shopping_cart_badge).to_have_text("1")

    def open_cart(self):
        with allure.step("Open shopping cart"):
            self.shopping_cart_link.click()